File: packages/dart/test/verify_test.dart
import 'package:qsu/qsu.dart';
import 'package:test/test.dart' hide contains, isEmpty;

void main() {
  group('Verify', () {
    test('isObject', () {
      expect(isObject('{}'), false);
      expect(isObject(true), false);
      expect(isObject(false), false);
      expect(isObject(null), false);
      expect(isObject(1), false);
      expect(isObject([]), false);
      expect(isObject(() => '123'), false);
      expect(isObject({}), true);
      expect(isObject([1, 2]), false);
      expect(
          isObject([
            {'a': 1, 'b': 2}
          ]),
          false);
      expect(isObject({'a': 1, 'b': 2}), true);
      expect(isObject({'a': {}, 'b': []}), true);
    });

    test('isEqual', () {
      const val1 = 'abc';
      const val2 = 'abc';
      const val3 = 'abc';

      expect(isEqual(1, [1, 2, 3]), false);
      expect(isEqual('abc', [val1, val2, val3]), true);
      expect(isEqual('123', ['123', 123]), true);
      expect(isEqual('123', ['123', 123, 123, 123]), true);
      expect(isEqual(123, '123'), true);
    });

    test('isEqualStrict', () {
      const val1 = 'abc';
      const val2 = 'abc';
      const val3 = 'abc';

      expect(isEqualStrict(1, [1, 2, 3, 4, 5]), false);
      expect(isEqualStrict('abc', [val1, val2, val3]), true);
      expect(isEqualStrict('123', ['123', 123]), false);
      expect(isEqualStrict('123', ['123', '123']), true);
      expect(isEqualStrict(123, '123'), false);
    });

    test('isEmpty', () {
      expect(isEmpty(''), true);
      expect(isEmpty('1234'), false);
      expect(isEmpty(1234), false);
      expect(isEmpty(1.234), false);
      expect(isEmpty(null), true);
      expect(isEmpty([]), true);
      expect(isEmpty([{}]), false);
      expect(isEmpty([[]]), false);
      expect(isEmpty(['1234']), false);
      expect(isEmpty({}), true);
      expect(isEmpty({'a': '1234'}), false);
    });

    test('contains', () {
      expect(contains('12345', '3'), true);
      expect(contains('12345', '10'), false);
      expect(contains('ABC', ['A', 'B', 'C']), true);
      expect(contains('ABC', ['D', 'E', 'F']), false);
      expect(contains('ABC', ['AB', 'C'], exact: true), false);
      expect(contains('AB', ['AB', 'C', 'D'], exact: true), true);
    });

    test('isUrl', () {
      expect(isUrl(''), false);
      expect(isUrl('https://'), false);
      expect(isUrl('www.google.com'), false);
      expect(isUrl('www.google.com', withProtocol: true), true);
      expect(isUrl('https://google.com'), true);
      expect(isUrl('https://google.com', withProtocol: true), true);
      expect(isUrl('https://google'), true);
      expect(isUrl('https://google', withProtocol: false, strict: true), false);
      expect(isUrl('https://google.com?query=qsu'), true);
    });

    test('is2dArray', () {
      expect(is2dArray([]), false);
      expect(is2dArray([[], []]), true);
      expect(
          is2dArray([
            {'a': 1},
            {'b': 2}
          ]),
          false);
      expect(
          is2dArray([
            [1],
            [2]
          ]),
          true);
    });

    test('isEmail', () {
      expect(isEmail('1@1.com'), true);
      expect(isEmail('abc@def.ghi'), true);
      expect(isEmail('Abc@def.ghi', onlyLowerCase: true), false);
      expect(isEmail('Abc@def.ghi'), true);
      expect(isEmail('abc@Def.ghi'), true);
      expect(isEmail('abc@def.Ghi'), true);
      expect(isEmail('ABC@DEF.GHI'), true);
      expect(isEmail('abc@sub.domain.com'), true);
      expect(isEmail('a.bc@d.ef'), true);
      expect(isEmail('a-12_34@b-12-34.net'), true);
      expect(isEmail('@b1234.net'), false);
      expect(isEmail('a1234@b1234'), false);
      expect(isEmail('a_1234@b_1234.net'), false);
      expect(isEmail('abc@@def.com'), false);
      expect(isEmail('11.com'), false);
      expect(isEmail('sub.domain.com'), false);
      expect(isEmail('1@1@a.com'), false);
    });

    test('between', () {
      expect(between([1, 10], 1), false);
      expect(between([1, 10], 1, inclusive: true), true);
      expect(between([10, 100], 11), true);
    });

    test('len', () {
      expect(len('12345'), 5);
      expect(len(12345), 5);
      expect(len(() => '123'), 3);
      expect(len([1, 2, 3, 4]), 4);
      expect(len({'hello': 'world', 'lorem': 'ipsum'}), 2);
      expect(
          len([
            {'hello': 1, 'world': 2},
            {'lorem': 3}
          ]),
          2);
    });

    test('isTrueMinimumNumberOfTimes', () {
      final int left = 2;
      final int right1 = 1 + 1;
      final int right2 = 2 + 1;

      expect(isTrueMinimumNumberOfTimes([true, false, false]), true);
      expect(isTrueMinimumNumberOfTimes([true, true], minimumCount: 1), true);
      expect(isTrueMinimumNumberOfTimes([true, false, true], minimumCount: 2),
          true);
      expect(isTrueMinimumNumberOfTimes([true, false, true], minimumCount: 1),
          true);
      expect(
          isTrueMinimumNumberOfTimes([left == right1, false, true, true, false],
              minimumCount: 3),
          true);
      expect(
          isTrueMinimumNumberOfTimes([left == right2, false, true, true, false],
              minimumCount: 3),
          false);
    });

    test('hasBadWords', () {
      const List<String> words = ['admin', 'apple'];

      expect(hasBadWords('', words: words), false);
      expect(hasBadWords('hello world', words: words), false);
      expect(hasBadWords('admin'), false);
      expect(hasBadWords('admin', words: []), false);
      expect(hasBadWords('admin', words: ['', '  ']), false);
      expect(hasBadWords('!!! ??? ***', words: words), false);

      expect(hasBadWords('i am admin', words: words), true);
      expect(hasBadWords('I AM ADMIN', words: words), true);
      expect(hasBadWords('pineapple juice', words: words), true);
      expect(hasBadWords('apple, banana', words: words), true);

      // Separators hidden inside the word.
      expect(hasBadWords('ad___min', words: words), true);
      expect(hasBadWords('a.d.m.i.n', words: words), true);
      expect(hasBadWords('ad\$min', words: words), true);
      expect(hasBadWords('a d m i n', words: words), true);
      expect(hasBadWords('the ad min account', words: words), true);

      // Digits wedged between the letters.
      expect(hasBadWords('ad1min', words: words), true);
      expect(hasBadWords('ap123ple', words: words), true);
      expect(hasBadWords('a1d.m2in', words: words), true);

      // Lookalike characters.
      expect(hasBadWords('adm1n', words: words), true);
      expect(hasBadWords('@dm1n', words: words), true);
      expect(hasBadWords('4pp13', words: words), true);
      expect(hasBadWords('ａｄｍｉｎ', words: words), true);
      expect(hasBadWords('аdmin', words: words), true); // Cyrillic 'а'
      expect(hasBadWords('ádmín', words: words), true);
      expect(hasBadWords('gㅇㅇd', words: ['good']), true); // 'ㅇ' as 'o'

      // Stylized Unicode letters.
      expect(hasBadWords('𝗮𝗱𝗺𝗶𝗻', words: words), true);
      expect(hasBadWords('ⓐⓓⓜⓘⓝ', words: words), true);
      expect(hasBadWords('app/e', words: words), true);

      // Allowed words are excused, even though they contain a banned word.
      expect(
          hasBadWords('pineapple juice',
              words: words, allowWords: ['pineapple']),
          false);
      expect(
          hasBadWords('apple and pineapple',
              words: words, allowWords: ['pineapple']),
          true);
      expect(
          hasBadWords('administrator here',
              words: words, allowWords: ['administrator']),
          false);

      // A word split over a space only counts from the start of a token.
      expect(hasBadWords('read min please', words: words), false);
      expect(hasBadWords('nomad mineral', words: words), false);

      const List<String> koWords = ['사과', '고양이'];

      expect(hasBadWords('맛있는 사과!', words: koWords), true);
      expect(hasBadWords('사과나무', words: koWords), true);
      expect(hasBadWords('사-과', words: koWords), true);
      expect(hasBadWords('사 과', words: koWords), true);
      expect(hasBadWords('우리 고양이는 귀엽다', words: koWords), true);

      // Decomposed jamo, including compound vowels and finals.
      expect(hasBadWords('ㅅㅏㄱㅗㅏ', words: koWords), true);
      expect(hasBadWords('사ㄱㅗㅏ', words: koWords), true);
      expect(hasBadWords('ㄱㅗㅇㅑㅇㅇl', words: koWords), true);
      expect(hasBadWords('ㅅr과', words: koWords), true); // 'r' shaped like 'ㅏ'

      // Digits wedged between the syllables.
      expect(hasBadWords('사1과', words: koWords), true);
      expect(hasBadWords('사123과', words: koWords), true);
      expect(hasBadWords('ㅅㅏ1ㄱㅗㅏ', words: koWords), true);
      expect(hasBadWords('고3양2이', words: koWords), true);

      // Unrelated words that only touch when the space is removed.
      expect(hasBadWords('이거사 과일이야', words: koWords), false);
      expect(hasBadWords('이거1사 과일이야', words: koWords), false);
      expect(hasBadWords('명사 과제', words: koWords), false);
      expect(hasBadWords('참고 양이 되었다', words: koWords), false);
    });
  });
}

File: packages/javascript/test/verify.test.ts
import assert from 'assert';
import { describe, it } from 'node:test';
import {
	isObject,
	isEqual,
	isEqualStrict,
	isEmpty,
	isUrl,
	contains,
	hasBadWords,
	is2dArray,
	between,
	len,
	isEmail,
	isTrueMinimumNumberOfTimes
} from '../dist';

describe('Verify', () => {
	it('isObject', () => {
		assert.strictEqual(isObject('{}'), false);
		assert.strictEqual(isObject(true), false);
		assert.strictEqual(isObject(false), false);
		assert.strictEqual(isObject(null), false);
		assert.strictEqual(isObject(undefined), false);
		assert.strictEqual(isObject(1), false);
		assert.strictEqual(isObject([]), false);
		assert.strictEqual(
			isObject(() => '123'),
			false
		);
		assert.strictEqual(isObject({}), true);
		assert.strictEqual(isObject([1, 2]), false);
		assert.strictEqual(isObject([{ a: 1, b: 2 }]), false);
		assert.strictEqual(isObject({ a: 1, b: 2 }), true);
		assert.strictEqual(isObject({ a: {}, b: [] }), true);
	});

	it('isEqual', () => {
		const val1 = 'abc';
		const val2 = 'abc';
		const val3 = 'abc';

		assert.strictEqual(isEqual(1, [1, 2, 3]), false);
		assert.strictEqual(isEqual('abc', [val1, val2, val3]), true);
		assert.strictEqual(isEqual('123', ['123', 123]), true);
		assert.strictEqual(isEqual(123, '123', 123), true);
	});

	it('isEqualStrict', () => {
		const val1 = 'abc';
		const val2 = 'abc';
		const val3 = 'abc';

		assert.strictEqual(isEqualStrict(1, [1, 2, 3, 4, 5]), false);
		assert.strictEqual(isEqualStrict('abc', [val1, val2, val3]), true);
		assert.strictEqual(isEqualStrict('123', ['123', 123]), false);
		assert.strictEqual(isEqualStrict('123', ['123', '123']), true);
		assert.strictEqual(isEqualStrict(123, '123', 123), false);
	});

	it('isEmpty', () => {
		assert.strictEqual(isEmpty(''), true);
		assert.strictEqual(isEmpty('1234'), false);
		assert.strictEqual(isEmpty(1234), false);
		assert.strictEqual(isEmpty(1.234), false);
		assert.strictEqual(isEmpty(null), true);
		assert.strictEqual(isEmpty([]), true);
		assert.strictEqual(isEmpty([{}]), false);
		assert.strictEqual(isEmpty([[]]), false);
		assert.strictEqual(isEmpty(['1234']), false);
		assert.strictEqual(isEmpty({}), true);
		assert.strictEqual(isEmpty({ a: '1234' }), false);
	});

	it('isUrl', () => {
		assert.strictEqual(isUrl(''), false);
		assert.strictEqual(isUrl('https://'), false);
		assert.strictEqual(isUrl('www.google.com'), false);
		assert.strictEqual(isUrl('www.google.com', true), true);
		assert.strictEqual(isUrl('https://google.com'), true);
		assert.strictEqual(isUrl('https://google.com', true), true);
		assert.strictEqual(isUrl('https://google'), true);
		assert.strictEqual(isUrl('https://google', false, true), false);
		assert.strictEqual(isUrl('https://google.com?query=qsu'), true);
	});

	it('contains', () => {
		assert.strictEqual(contains('12345', '3'), true);
		assert.strictEqual(contains('12345', '10'), false);
		assert.strictEqual(contains('ABC', ['A', 'B', 'C']), true);
		assert.strictEqual(contains('ABC', ['D', 'E', 'F']), false);
		assert.strictEqual(contains('ABC', ['AB', 'C'], true), false);
		assert.strictEqual(contains('AB', ['AB', 'C', 'D'], true), true);
	});

	it('is2dArray', () => {
		assert.strictEqual(is2dArray([]), false);
		assert.strictEqual(is2dArray([[], []]), true);
		assert.strictEqual(is2dArray([{ a: 1 }, { b: 2 }]), false);
		assert.strictEqual(is2dArray([[1], [2]]), true);
	});

	it('between', () => {
		assert.strictEqual(between([1, 10], 1), false);
		assert.strictEqual(between([1, 10], 1, true), true);
		assert.strictEqual(between([10, 100], 11), true);
	});

	it('len', () => {
		assert.strictEqual(len('12345'), 5);
		assert.strictEqual(len(12345), 5);
		assert.strictEqual(
			len(() => '123'),
			3
		);
		assert.strictEqual(len([1, 2, 3, 4]), 4);
		assert.strictEqual(len({ hello: 'world', lorem: 'ipsum' }), 2);
		assert.strictEqual(len([{ hello: 1, world: 2 }, { lorem: 3 }]), 2);
	});

	it('isEmail', () => {
		assert.strictEqual(isEmail('1@1.com'), true);
		assert.strictEqual(isEmail('abc@def.ghi'), true);
		assert.strictEqual(isEmail('Abc@def.ghi', true), false); // Case-sensitive
		assert.strictEqual(isEmail('Abc@def.ghi'), true);
		assert.strictEqual(isEmail('abc@Def.ghi'), true);
		assert.strictEqual(isEmail('abc@def.Ghi'), true);
		assert.strictEqual(isEmail('ABC@DEF.GHI'), true);
		assert.strictEqual(isEmail('abc@sub.domain.com'), true);
		assert.strictEqual(isEmail('a.bc@d.ef'), true);
		assert.strictEqual(isEmail('a-12_34@b-12-34.net'), true);
		assert.strictEqual(isEmail('@b1234.net'), false);
		assert.strictEqual(isEmail('a1234@b1234'), false);
		assert.strictEqual(isEmail('a_1234@b_1234.net'), false);
		assert.strictEqual(isEmail('abc@@def.com'), false);
		assert.strictEqual(isEmail('11.com'), false);
		assert.strictEqual(isEmail('sub.domain.com'), false);
		assert.strictEqual(isEmail('1@1@a.com'), false);
		assert.strictEqual(isEmail('a.com@a'), false);
	});

	it('isTrueMinimumNumberOfTimes', () => {
		const left = 2;
		const right1 = 1 + 1;
		const right2 = 2 + 1;

		assert.strictEqual(isTrueMinimumNumberOfTimes([true, false, false]), true);
		assert.strictEqual(isTrueMinimumNumberOfTimes([true, true], 1), true);
		assert.strictEqual(isTrueMinimumNumberOfTimes([true, false, true], 2), true);
		assert.strictEqual(isTrueMinimumNumberOfTimes([true, false, true], 1), true);
		assert.strictEqual(
			isTrueMinimumNumberOfTimes([left === right1, false, true, true, false], 3),
			true
		);
		assert.strictEqual(
			isTrueMinimumNumberOfTimes([left === right2, false, true, true, false], 3),
			false
		);
	});

	it('hasBadWords', () => {
		const words = ['admin', 'apple'];

		assert.strictEqual(hasBadWords('', words), false);
		assert.strictEqual(hasBadWords('hello world', words), false);
		assert.strictEqual(hasBadWords('admin'), false);
		assert.strictEqual(hasBadWords('admin', []), false);
		assert.strictEqual(hasBadWords('admin', ['', '  ']), false);
		assert.strictEqual(hasBadWords('!!! ??? ***', words), false);

		assert.strictEqual(hasBadWords('i am admin', words), true);
		assert.strictEqual(hasBadWords('I AM ADMIN', words), true);
		assert.strictEqual(hasBadWords('pineapple juice', words), true);
		assert.strictEqual(hasBadWords('apple, banana', words), true);

		// Separators hidden inside the word.
		assert.strictEqual(hasBadWords('ad___min', words), true);
		assert.strictEqual(hasBadWords('a.d.m.i.n', words), true);
		assert.strictEqual(hasBadWords('ad$min', words), true);
		assert.strictEqual(hasBadWords('a d m i n', words), true);
		assert.strictEqual(hasBadWords('the ad min account', words), true);

		// Digits wedged between the letters.
		assert.strictEqual(hasBadWords('ad1min', words), true);
		assert.strictEqual(hasBadWords('ap123ple', words), true);
		assert.strictEqual(hasBadWords('a1d.m2in', words), true);

		// Lookalike characters.
		assert.strictEqual(hasBadWords('adm1n', words), true);
		assert.strictEqual(hasBadWords('@dm1n', words), true);
		assert.strictEqual(hasBadWords('4pp13', words), true);
		assert.strictEqual(hasBadWords('ａｄｍｉｎ', words), true);
		assert.strictEqual(hasBadWords('аdmin', words), true); // Cyrillic 'а'
		assert.strictEqual(hasBadWords('ádmín', words), true);
		assert.strictEqual(hasBadWords('gㅇㅇd', ['good']), true); // Hangul 'ㅇ' as 'o'

		// Stylized Unicode letters.
		assert.strictEqual(hasBadWords('𝗮𝗱𝗺𝗶𝗻', words), true);
		assert.strictEqual(hasBadWords('ⓐⓓⓜⓘⓝ', words), true);
		assert.strictEqual(hasBadWords('app/e', words), true);

		// Allowed words are excused, even though they contain a banned word.
		assert.strictEqual(hasBadWords('pineapple juice', words, ['pineapple']), false);
		assert.strictEqual(hasBadWords('apple and pineapple', words, ['pineapple']), true);
		assert.strictEqual(hasBadWords('administrator here', words, ['administrator']), false);

		// A word split over a space only counts from the start of a token.
		assert.strictEqual(hasBadWords('read min please', words), false);
		assert.strictEqual(hasBadWords('nomad mineral', words), false);

		const koWords = ['사과', '고양이'];

		assert.strictEqual(hasBadWords('맛있는 사과!', koWords), true);
		assert.strictEqual(hasBadWords('사과나무', koWords), true);
		assert.strictEqual(hasBadWords('사-과', koWords), true);
		assert.strictEqual(hasBadWords('사 과', koWords), true);
		assert.strictEqual(hasBadWords('우리 고양이는 귀엽다', koWords), true);

		// Decomposed jamo, including compound vowels and finals.
		assert.strictEqual(hasBadWords('ㅅㅏㄱㅗㅏ', koWords), true);
		assert.strictEqual(hasBadWords('사ㄱㅗㅏ', koWords), true);
		assert.strictEqual(hasBadWords('ㄱㅗㅇㅑㅇㅇl', koWords), true);
		assert.strictEqual(hasBadWords('ㅅr과', koWords), true); // 'r' shaped like 'ㅏ'

		// Digits wedged between the syllables.
		assert.strictEqual(hasBadWords('사1과', koWords), true);
		assert.strictEqual(hasBadWords('사123과', koWords), true);
		assert.strictEqual(hasBadWords('ㅅㅏ1ㄱㅗㅏ', koWords), true);
		assert.strictEqual(hasBadWords('고3양2이', koWords), true);

		// Unrelated words that only touch when the space is removed.
		assert.strictEqual(hasBadWords('이거사 과일이야', koWords), false);
		assert.strictEqual(hasBadWords('이거1사 과일이야', koWords), false);
		assert.strictEqual(hasBadWords('명사 과제', koWords), false);
		assert.strictEqual(hasBadWords('참고 양이 되었다', koWords), false);
	});
});
